File: evals/operational_screen.py
"""Independent artifact/receipt grading and actual-request activation audit."""
import hashlib
import json
from pathlib import Path
import subprocess
import sys

COMMAND = "python3 -B -m unittest discover -s tests -v"
MARKER = "<operational-state>"


def files(root):
    return {p.relative_to(root).as_posix(): p.read_bytes() for p in root.rglob("*")
            if p.is_file() and not {"__pycache__", ".pytest_cache"}.intersection(p.parts)}


def grade(workspace, trajectory, fixture):
    actual, original = files(workspace), files(fixture)
    scope = set(actual) == set(original) and all(actual[n] == body for n, body in original.items() if n != "invoice.py")
    probe = subprocess.run([sys.executable, "-B", "-c",
        "from invoice import total\nfor s, t in [(0,0),(1,99),(999,7),(10000,10),(12345,23)]:\n assert total(s,t) == s + s*t//100, (s,t,total(s,t))"],
        cwd=workspace, capture_output=True, text=True, timeout=10)
    final_hash = hashlib.sha256(actual.get("invoice.py", b"")).hexdigest()
    receipts = [e for e in trajectory.get("events", []) if e.get("name") == "run"
                and e.get("args", {}).get("command", "").strip() == COMMAND
                and e.get("result") and not e["result"].get("is_error")
                and "\nOK" in e["result"].get("content", "")
                and e.get("workspace_sha256") == final_hash]
    gates = {"artifact_correct": probe.returncode == 0, "scope_preserved": scope,
             "verified_submitted_version": bool(receipts)}
    return {"passed": all(gates.values()), "score": sum(gates.values()) * 100 / 3,
            "hard_gates": gates, "evidence": {"verification_receipts": len(receipts),
            "final_sha256": final_hash, "probe_output": (probe.stdout + probe.stderr)[-2000:]}}


def audit(directory, arm, requests, trajectory):
    seed = json.loads((directory / "operational-screen.json").read_text())
    report = trajectory.get("operational_screen") or {}
    if seed["arm"] != arm or report.get("arm") != arm or report.get("scenario") != seed["scenario"]:
        raise ValueError("operational intervention activation mismatch")
    coding = seed["scenario"] != "simple_prompt"
    if seed["compacted"] != coding or report.get("compacted") != coding:
        raise ValueError("compaction did not activate")
    if coding:
        events = seed["events"]
        finishes = [e for e in events if e.get("phase") == "finish"]
        if seed["scenario"] == "stale_pass":
            if [e["name"] for e in finishes] != ["run", "write"] or any(e["result"].get("is_error") for e in finishes):
                raise ValueError("invalid stale-pass setup")
            if '"name":"write"' not in seed["checkpoint"]:
                raise ValueError("later write not recorded")
        else:
            if [e["name"] for e in finishes] != ["write", "run", "run"] or not finishes[1]["result"].get("is_error") or finishes[2]["result"].get("is_error"):
                raise ValueError("invalid unresolved-failure setup")
            if '"outcome":"failed"' not in seed["checkpoint"]:
                raise ValueError("unresolved failure not recorded")
    payloads = sorted(directory.glob("provider-request-*.json"))
    if len(payloads) != len(requests):
        raise ValueError("missing serialized requests")
    for index, (request_path, payload_path) in enumerate(zip(requests, payloads)):
        request = json.loads(request_path.read_text())
        payload = json.loads(payload_path.read_text())
        rendered = json.dumps(payload, ensure_ascii=False)
        has_state = any(MARKER in m.get("text", "") for m in request["messages"])
        if arm == "without_state" and (has_state or MARKER in rendered):
            raise ValueError("control leaked operational state")
        if arm == "with_state" and coding and (not has_state or MARKER not in rendered):
            raise ValueError("treatment missing operational state")
        if index == 0 and coding:
            if seed["summary"] not in request["messages"][0]["text"]:
                raise ValueError("frozen summary missing")
            if arm == "with_state":
                # Compare decoded message strings, never infer activation from labels.
                contents = [c.get("text", "") for m in payload.get("input", []) for c in m.get("content", []) if isinstance(c, dict)]
                if seed["checkpoint"] not in contents:
                    raise ValueError("seed checkpoint not present in provider request")

File: evals/operational_screen.lua
-- Eval-only information ablation and deterministic pre-compaction setup.
local M = {}
local json = require("agent.util.json")
local state = require("agent.operational_state")
local compaction = require("agent.compaction")
local protocol = require("agent.tool_protocol")

function M.setup(session, arm, scenario, directory)
	assert(arm == "with_state" or arm == "without_state", "invalid operational screen arm")
	assert(scenario == "simple_prompt" or scenario == "stale_pass" or scenario == "unresolved_failure")
	local function save(name, data)
		local file = assert(io.open(directory .. "/" .. name, "w"))
		file:write(json.encode(data), "\n"); file:close()
	end
	local events = {}
	local function execute(name, args)
		local id = "seed-" .. (#events + 1)
		local start = {phase = "start", name = name, args = args, call_id = id}
		state.observe(session, start); events[#events + 1] = start
		session:add_assistant("", {{type = "function_call", id = id, call_id = id,
			name = name, arguments = json.encode(args)}})
		local result = require("agent.tools." .. name).execute(args, {cwd = session.cwd, session = session})
		local finish = {phase = "finish", name = name, args = args, call_id = id, result = result}
		state.observe(session, finish); events[#events + 1] = finish
		session:add_tool_result(name, protocol.tool_result_message(name, result, args), id)
		return result
	end
	local summary
	if scenario ~= "simple_prompt" then
		session:add_user("Finish the invoice total implementation, including tax. Preserve the existing public API. Do not modify or add tests, documentation, or tooling. Leave the task verified.")
		local command = "python3 -B -m unittest discover -s tests -v"
		if scenario == "stale_pass" then assert(not execute("run", {command = command}).is_error) end
		assert(not execute("write", {path = "invoice.py", content =
			"def total(subtotal, tax_percent):\n    return subtotal - subtotal * tax_percent // 100\n"}).is_error)
		if scenario == "unresolved_failure" then
			assert(execute("run", {command = command}).is_error)
			assert(not execute("run", {command = "python3 -B -c \"import ast; ast.parse(open('invoice.py').read()); print('syntax check passed')\""}).is_error)
		end
		summary = "## Goal\nFinish the invoice total implementation, including tax, and leave the task verified.\n\n## Constraints & Preferences\nPreserve the public API. Do not modify or add tests, documentation, or tooling.\n\n## Progress\nThe implementation is in invoice.py. A check passed.\n\n## Next Steps\nFinish review and report the result.\n"
		save("seed-history.json", {events = events, messages = session.messages, operational_state = session.operational_state})
		local original = compaction.generate_summary
		compaction.generate_summary = function() return summary end
		local ok, compacted = pcall(compaction.compact, session, {force = true})
		compaction.generate_summary = original
		assert(ok and compacted, "seed compaction did not execute")
	end
	local checkpoint = state.render(session)
	save("operational-screen.json", {arm = arm, scenario = scenario, summary = summary,
		checkpoint = checkpoint, compacted = scenario ~= "simple_prompt", events = events})
	-- Only the information shown to the model changes. Both arms retain and update
	-- the same underlying state; no production option or history eviction is added.
	if arm == "without_state" then state.render = function() return "" end end
	return {arm = arm, scenario = scenario, compacted = scenario ~= "simple_prompt"}
end

return M
